proof/refine/ARM/Ipc_R.thy

```diff
@@ -2522,18 +2522,499 @@ global_interpretation setMessageInfo: typ_at_all_props' "setMessageInfo t info"
+    apply (fastforce elim: sc_at_cross tcb_at_cross simp: state_relation_def)
+   apply clarsimp
+   apply (frule valid_objs'_valid_tcbs')
+   apply (rule valid_objsE', assumption)
+    apply (fastforce simp: obj_at'_def projectKO_eq projectKO_sc)
+   apply (clarsimp simp: valid_obj'_def valid_sched_context'_def obj_at'_def projectKOs)
+   apply (frule valid_objs'_valid_tcbs')
+   apply (fastforce simp: valid_obj'_def valid_tcb'_def)
+  apply (clarsimp simp: sym_refs_asrt_def)
+  done
+
+lemma maybeDonateSc_corres:
+  "corres dc (tcb_at tcb_ptr and ntfn_at ntfn_ptr and invs and weak_valid_sched_action)
+             (tcb_at' tcb_ptr and ntfn_at' ntfn_ptr and (\<lambda>s. valid_objs' s \<and>
+                              valid_queues s \<and>
+                              valid_queues' s \<and>
+                              valid_release_queue s \<and>
+                              valid_release_queue' s))
+            (maybe_donate_sc tcb_ptr ntfn_ptr)
+            (maybeDonateSc tcb_ptr ntfn_ptr)"
+  unfolding maybeDonateSc_def maybe_donate_sc_def
+  apply (simp add: get_tcb_obj_ref_def get_ntfn_obj_ref_def2 liftM_def maybeM_def
+                   get_sc_obj_ref_def2)
+  apply (rule corres_guard_imp)
+    apply (rule corres_split[OF _ threadget_corres, where r'="(=)"])
+       apply (rule corres_when, simp)
+       apply (rule corres_split[OF _ get_ntfn_corres])
+         apply (rule corres_option_split)
+           apply (clarsimp simp: ntfn_relation_def)
+          apply (rule corres_return_trivial)
+         apply (simp add: get_tcb_obj_ref_def get_ntfn_obj_ref_def2 liftM_def maybeM_def
+                          get_sc_obj_ref_def2)
+         apply (rule corres_split[OF _ get_sc_corres])
+           apply (rule corres_when)
+            apply (clarsimp simp: sc_relation_def)
+           apply (rule corres_split[OF _ schedContextDonate_corres])
+             apply (rule corres_split[OF _ getCurSc_corres])
+               apply (rule corres_split[OF _ corres_when])
+                   apply (rule schedContextResume_corres, simp)
+                  apply clarsimp
+                 apply (rule refillUnblockCheck_corres)
+                apply (wpsimp wp: get_simple_ko_wp getNotification_wp)+
+      apply (clarsimp simp: tcb_relation_def)
+     apply (wpsimp wp: thread_get_wp' threadGet_wp)+
+   apply (clarsimp split: option.splits)
+   apply (frule (1) valid_objs_ko_at[OF invs_valid_objs, where ptr=ntfn_ptr])
+   apply (clarsimp simp: valid_obj_def valid_ntfn_def)
+  apply (clarsimp simp: tcb_at'_ex_eq_all split: option.splits)
+  apply (frule (1) ntfn_ko_at_valid_objs_valid_ntfn')
+  apply (clarsimp simp: valid_ntfn'_def)
+  done
+
+crunches refillUnblockCheck
+  for valid_release_queue[wp]: valid_release_queue
+  and valid_release_queue'[wp]: valid_release_queue'
+  and valid_queues[wp]: valid_queues
+  and valid_queues'[wp]: valid_queues'
+  (wp: whileM_inv crunch_wps)
+
+lemma threadSet_valid_release_queue_inv:
```

Comment:
This one doesn’t seem to be used. Maybe we don’t need this one, if we have the one below? (And rename the one below?)

Reply:
I think it's a great idea to rename them both. I think the first one will be needed by someone at some point, it says something useful and if it's ok with you I'd like to leave it in.